sockeye/scoring.py

```diff
@@ -0,0 +1,474 @@
+    # print("sources {} source vocabs {}".format(sources, source_vocabs))
+    sources_sentences = [data_io.SequenceReader(source, vocab, add_bos=False) for source, vocab in zip(sources, source_vocabs)]
+    target_sentences = data_io.SequenceReader(args.target, target_vocab, add_bos=True, limit=None)
+
+    # Pass 1: get target/source length ratios.
+    length_statistics = data_io.analyze_sequence_lengths(sources, args.target, source_vocabs, target_vocab, max_seq_len_source, max_seq_len_target)
+
+    # define buckets
+    bucketing = not args.no_bucketing
+    buckets = data_io.define_parallel_buckets(max_seq_len_source, max_seq_len_target, args.bucket_width, length_statistics.length_ratio_mean) if bucketing else [(max_seq_len_source, max_seq_len_target)]
+
+    # get iterator
+    # Pass 2: Get data statistics
+    data_statistics = data_io.get_data_statistics(sources_sentences,
+                                                  target_sentences,
+                                                  buckets,
+                                                  length_statistics.length_ratio_mean,
+                                                  length_statistics.length_ratio_std,
+                                                  source_vocabs,
+                                                  target_vocab)
+
+    bucket_batch_sizes = data_io.define_bucket_batch_sizes(buckets,
+                                                           args.batch_size,
+                                                           batch_by_words,
+                                                           batch_num_devices,
+                                                           data_statistics.average_len_target_per_bucket)
+
+    data_statistics.log(bucket_batch_sizes)
+
+    data_loader = data_io.RawParallelDatasetLoader(buckets=buckets,
+                                                   eos_id=target_vocab[C.EOS_SYMBOL],
+                                                   pad_id=C.PAD_ID,
+                                                   skip_if_no_bucket=False)
+
+    parallel_data = data_loader.load(sources_sentences, target_sentences,
+                                     data_statistics.num_sents_per_bucket).fill_up(bucket_batch_sizes, args.fill_up)
+    map_buckets2sentence_ids = data_loader.map_buckets2sentence_ids
+
+    data_iter = data_io.ParallelSampleIter(parallel_data,
+                                           buckets,
+                                           args.batch_size,
+                                           bucket_batch_sizes,
+                                           shuffle=False)
+
+    return data_iter, config_, map_buckets2sentence_ids
+
+
+def get_max_source_and_target(args: argparse.Namespace) -> Tuple[int, int]:
+    source_lines = utils.smart_open(args.source).readlines()
+    target_lines = utils.smart_open(args.target).readlines()
+    max_len_source = max([len(line.rstrip().split()) for line in source_lines])
```

Comment:
you could use `data_io.read_content` here and maybe a generator instead of constructing the full list:
```
max([len(sent) for sent in data_io.read_content(source))
```